### PR TITLE
extension: add ExtensionPlugin class

### PR DESCRIPTION
**Pull request by steweg**
This patch introduces new class ExtensionPlugin, which is wrapper around libyang extension plugin, which allows user to define custom action for parsing, compiling, and freeing parsed or compiled extensions.

Custom actions can also raise a new type of exception LibyangExtensionError, which allows proper translation of exception to libyang error codes and logging of error message

### CODE REVIEW

**Comment by samuel-gauthier** [comment]
@steweg, what do you want to do with this draft? Shall we close it or do you have questions?

**Comment by steweg** [comment]
I will need to rebase it and then I will mark it for review

File: libyang/extension.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        name: str,
+        id_str: str,
+        context: Optional[Context] = None,
+        parse_clb: Optional[Callable[[Module, ExtensionParsed], None]] = None,
+        compile_clb: Optional[
+            Callable[[ExtensionParsed, ExtensionCompiled], None]
+        ] = None,
+        parse_free_clb: Optional[Callable[[ExtensionParsed], None]] = None,
+        compile_free_clb: Optional[Callable[[ExtensionCompiled], None]] = None,
+    ) -> None:
```
Comment (samuel-gauthier) [inline libyang/extension.py:104]: We need a docstring to explain the arguments and callback formats.

File: libyang/schema.py [hunk 2]
```diff
@@ -411,6 +411,14 @@ def name(self) -> str:
     def module(self) -> Module:
         return self._module_from_parsed()
 
+    def parent_node(self) -> Optional["PNode"]:
```
Comment (samuel-gauthier) [inline libyang/schema.py:414]: Is this related to the extension plugin?

File: libyang/schema.py [hunk 3]
```diff
@@ -394,7 +394,7 @@ def __str__(self):
 class ExtensionParsed(Extension):
     __slots__ = ("module_parent",)
 
-    def __init__(self, context: "libyang.Context", cdata, module_parent: Module = None):
+    def __init__(self, context: "libyang.Context", cdata, module_parent: Module):
```
Comment (samuel-gauthier) [inline libyang/schema.py:397]: Is this related to the extension plugin?

File: libyang/schema.py [hunk 4]
```diff
@@ -428,6 +436,14 @@ def module(self) -> Module:
             raise self.context.error("cannot get module")
         return Module(self.context, self.cdata_def.module)
 
+    def parent_node(self) -> Optional["SNode"]:
```
Comment (samuel-gauthier) [inline libyang/schema.py:439]: Is this related to the extension plugin?

**Comment by samuel-gauthier** [comment]
Having a plugin that counts calls is fine for non-regression testing. Could we have an example of a meaningful plugin? That would help users to play with this feature. How about contributing to the examples section of the README file for this one?

File: libyang/extension.py (hunk 1 above)
Comment (steweg) [reply libyang/extension.py:104]: Done

Comment (steweg) [reply libyang/extension.py:104]: Done

File: libyang/schema.py (hunk 2 above)
Comment (steweg) [reply libyang/schema.py:414]: Yes, but indirectly. In real life, when you define your own extension plugin, on 90% of cases you would like to perform some validation whether extension is used correctly (e.g. is shall be used only on LEAF nodes etc..). This function is general way how to access the parent PNode to perform this.

File: libyang/schema.py (hunk 4 above)
Comment (steweg) [reply libyang/schema.py:439]: Yes, but indirectly. Same as for parsed extension explained above, this one is for compiled extension

File: libyang/schema.py (hunk 3 above)
Comment (steweg) [reply libyang/schema.py:397]: Yes, it turned out that extension parsed must have always module defined. Otherwise the rest of the code to get to `parent_node()` or `module_from_parsed()` doesn't really work

File: libyang/schema.py (hunk 2 above)
Comment (samuel-gauthier) [reply libyang/schema.py:414]: Then let's add a separate commit before this one adding support for parent_node for ExtensionParsed and ExtensionCompiled.

File: libyang/schema.py (hunk 3 above)
Comment (samuel-gauthier) [reply libyang/schema.py:397]: Two remarks:
- that's an api breakage, so it is not too nice
- this commit instantiates  an ExtensionParsed object with a None module in libyang_c_lyplg_ext_parse_free_clb, pointing out that there may be a need for it

So I would rather have _module_from_parsed take care of the case where the ExtensionParsed instance has no module, as we do in other parts of the code for cdata or cdata_parsed. Raising self.context.error("cannot get module") if self.module_parent is not set in module_from_parsed would do the trick, right?

Comment (steweg) [reply libyang/schema.py:397]: Fixed

File: libyang/extension.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                extension: The exact extension instance
+            Expected raises:
+                LibyangExtensionError in case of processing error
+        :arg compile_clb:
+            The optional callback function of which will be called during extension compiling
+            Expected arguments are:
+                extension_parsed: The parsed extension instance
+                extension_compiled: The compiled extension instance
+            Expected raises:
+                LibyangExtensionError in case of processing error
```
Comment (samuel-gauthier) [inline libyang/extension.py:133]: Is it not raising a LibyangError? That's what you test for in the unit test.

File: libyang/schema.py (hunk 2 above)
Comment (steweg) [reply libyang/schema.py:416]: Done

**Comment by samuel-gauthier** [comment]
@steweg there is this warning when running the test after the last commit of this series, can you check it?
Exception ignored from cffi callback <function libyang_c_lyplg_ext_parse_clb at 0x7ff3451ceaf0>:
Traceback (most recent call last):
  File "/home/runner/work/libyang-python/libyang-python/.tox/py39/lib/python3.9/site-packages/libyang/extension.py", line 35, in libyang_c_lyplg_ext_parse_clb
    plugin.parse_clb(module, parsed_ext)
  File "/home/runner/work/libyang-python/libyang-python/tests/test_extension.py", line 150, in _parse_clb
    self._verify_single(parent)
  File "/home/runner/work/libyang-python/libyang-python/tests/test_extension.py", line 132, in _verify_single
    if e.name() == self.name and e.module().name() == self.module_name:
  File "/home/runner/work/libyang-python/libyang-python/.tox/py39/lib/python3.9/site-packages/libyang/schema.py", line 414, in module
    return self._module_from_parsed()
  File "/home/runner/work/libyang-python/libyang-python/.tox/py39/lib/python3.9/site-packages/libyang/schema.py", line 408, in _module_from_parsed
    raise self.context.error("cannot get module")
libyang.util.LibyangError: cannot get module
.Exception ignored from cffi callback <function libyang_c_lyplg_ext_parse_clb at 0x7ff3451ceaf0>:
Traceback (most recent call last):
  File "/home/runner/work/libyang-python/libyang-python/.tox/py39/lib/python3.9/site-packages/libyang/extension.py", line 35, in libyang_c_lyplg_ext_parse_clb
    plugin.parse_clb(module, parsed_ext)
  File "/home/runner/work/libyang-python/libyang-python/tests/test_extension.py", line 150, in _parse_clb
    self._verify_single(parent)
  File "/home/runner/work/libyang-python/libyang-python/tests/test_extension.py", line 132, in _verify_single
    if e.name() == self.name and e.module().name() == self.module_name:
  File "/home/runner/work/libyang-python/libyang-python/.tox/py39/lib/python3.9/site-packages/libyang/schema.py", line 414, in module
    return self._module_from_parsed()
  File "/home/runner/work/libyang-python/libyang-python/.tox/py39/lib/python3.9/site-packages/libyang/schema.py", line 408, in _module_from_parsed
    raise self.context.error("cannot get module")
libyang.util.LibyangError: cannot get module

**Comment by steweg** [comment]
Fixed + rebased
